Code/Framework/AzCore/AzCore/DOM/DomPath.h

```diff
@@ -0,0 +1,132 @@
+    //! - Index, a numerical index for indexing within Arrays and Nodes
+    //! - Key, a name for indexing within Objects and Nodes
+    //! - EndOfArray, a special-case indicator for representing the end of an array
+    //!   used by the patching system to represent push / pop back operations.
+    class PathEntry final
+    {
+    public:
+        static constexpr size_t EndOfArrayIndex = size_t(-1);
+
+        PathEntry() = default;
+        PathEntry(const PathEntry&) = default;
+        PathEntry(PathEntry&&) = default;
+        explicit PathEntry(size_t value);
+        explicit PathEntry(AZ::Name value);
+        explicit PathEntry(AZStd::string_view value);
+
+        PathEntry& operator=(const PathEntry&) = default;
+        PathEntry& operator=(PathEntry&&) = default;
+        PathEntry& operator=(size_t value);
+        PathEntry& operator=(AZ::Name value);
+        PathEntry& operator=(AZStd::string_view value);
+
+        bool operator==(const PathEntry& other) const;
+        bool operator==(size_t index) const;
+        bool operator==(const AZ::Name& key) const;
+        bool operator==(AZStd::string_view key) const;
+        bool operator!=(const PathEntry& other) const;
+        bool operator!=(size_t index) const;
+        bool operator!=(const AZ::Name& key) const;
+        bool operator!=(AZStd::string_view key) const;
+
+        void SetEndOfArray();
+
+        bool IsEndOfArray() const;
+        bool IsIndex() const;
+        bool IsKey() const;
+
+        size_t GetIndex() const;
+        const AZ::Name& GetKey() const;
+
+    private:
+        AZStd::variant<size_t, AZ::Name> m_value;
+    };
+
+    //! Represents a path, represented as a series of PathEntry values, to a position in a Value.
+    class Path final
+    {
+    public:
+        using ContainerType = AZStd::vector<PathEntry>;
+        static constexpr char PathSeparator = '/';
+        static constexpr char EscapeCharacter = '~';
+        static constexpr char TildeSequence = '0';
+        static constexpr char ForwardSlashSequence = '1';
+        static constexpr char EndOfArrayCharacter = '-';
+
+        Path() = default;
+        Path(const Path&) = default;
+        Path(Path&&) = default;
+        explicit Path(AZStd::initializer_list<PathEntry> init);
+        //! Creates a Path from a path string, a path string is formatted per the JSON pointer specification
+        //! and looks like "/path/to/value/0"
+        explicit Path(AZStd::string_view pathString);
+
+        template<class InputIterator>
+        explicit Path(InputIterator first, InputIterator last)
+            : m_entries(first, last)
+        {
+        }
+
+        Path& operator=(const Path&) = default;
+        Path& operator=(Path&&) = default;
+
+        Path operator/(const PathEntry&) const;
+        Path operator/(size_t) const;
+        Path operator/(AZ::Name) const;
+        Path operator/(AZStd::string_view) const;
+        Path operator+(const Path&) const;
```

Comment:
I am not a fan of the `operator+` here.
It performs the same operation as the append operator `/`, while having a different symbol.

Plus developers who are familiar with an API with such as [std::filesystem::path](https://en.cppreference.com/w/cpp/filesystem/path#Concatenation), sees that the operator+ is used to perform string concatenation while `operator/` is used to append path segments

Reply:
Also this is missing the "append and assign combination operator" `operator/=`

Reply:
I don't think operator+ is doing the same thing as operator/ . Looking at the .cpp file, operator+ looks like it is in fact implemented as a concatenation operator.

Reply:
@srikappa-amzn is correct, I was modeling something similar to the `std` path class - I can remove `operator+` if desired though.

Reply:
It is by definition an append operation.
It is just append each all segments from the input parameter to the `*this` path and returning a copy.

What [concatenation](https://en.cppreference.com/w/cpp/filesystem/path/concat) means in terms of the `std::filesystem::path` is not what that function is doing.
The `std::filesystem::path::operator+=` is performing string concatenation of the underlying string types of the path without respect to any path separators or path segment replacement logic.

Reply:
The operations performed being append and concatenation are different.
i.e
```
std::filesystem::path path1("C:\\o3de");
std::filesystem::path path2("D:\\Users\\testuser");
std::filesystem::path pathConcat = path1 + path2; // result is "C:\\o3deD:\\Users\\testuser"
std::filesystem::path pathAppend = path1 / path2; // result is "D:\\Users\\testuser"
```

Reply:
Gotcha - do you think this would be clearer if I just used `operator/` or make it an explicit method? I don't think `std::filesystem::path` style concatenation will be useful here.

Reply:
I think it should be change to `operator/`, that way it is more inline with expectation that the second path is being append to the first path

Reply:
Done!